### ISSUE TITLE
Implementing solve() function in LinearAlgebra package targeting tridiagonal matrices

### ISSUE THREAD

**Issue by SamuelHoward** [chapel-lang/chapel#15338]
### Summary of Feature Request

Currently, the LinearAlgebra package has a general `solve()` implementation to solve the linear system 'A * x = b' in the general case, as well as `solve_triu()` and `solve_tril()` for the special cases where A is an upper or lower triangular matrix respectively. I think it would be beneficial to implement a function `solve_triag()` that would solve 'A * x = b' given that A is a tridiagonal matrix.  

I believe other languages use tridiagonal matrices when performing algorithms relating to eigenvalues amongst other things. [ScaLAPACK seems to utilize tridiagonal matrices](http://www.netlib.org/scalapack/slug/node81.html#SECTION04440000000000000000), although I could not find details on exactly how they are utilized. This makes me think `solve_triag()` would be worth implementing.

A somewhat simple algorithm for solving such systems can be found [here](https://www.cfd-online.com/Wiki/Tridiagonal_matrix_algorithm_-_TDMA_(Thomas_algorithm)). I also believe this would make progress towards bringing Chapel's LinearAlgebra module closer to Python's scipy.linalg. Specifically `solve_triag()` could make implementing [scipy's `solve_banded()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve_banded.html#scipy.linalg.solve_banded) easier, since a tridiagonal matrix is a special case of a banded matrix.

I believe this issue is most relevant to @ben-albrecht . Thank you!

**Comment by ben-albrecht** [comment]
@SamuelHoward - it would be beneficial to support both an implementation that wraps the appropriate LAPACK routine and a native Chapel implementation. I think it would wise to first implement the LAPACK-backed version, as that should be significantly less effort to get a working implementation. The `scipy` implementation and tests can serve as some guidance for this. 

That said, you are welcome to take a stab at the Chapel-native implementation first if that is your preference/interest.

Design-wise, what is interface do you propose e.g. proc name, arguments, return type?

**Comment by SamuelHoward** [comment]
I agree with what you said about implementing a LAPACK-backed version first.

As for the interface:

I was thinking the proc name could be `solve_tridiag()`, following a similar naming scheme to `solve_triu()` and `solve_tril()`. 

The arguments should be similar to the arguments of `solve()`, as we essentially still need the matrix A and the vector b. However, allowing A to be the tridiagonal matrix unchanged may be a mistake in terms of memory efficiency. 

Since A is tridiagonal, most of the entries of A are known to be 0. If A is an n * n matrix, there are n elements making up the diagonal and n - 1 elements making up the superdiagonal and subdiagonal of the matrix. I suggest storing these groups together in an array and passing that as an argument rather than passing A itself. [Scipy does this using 'matrix diagonal ordered form' for storing banded matrices.](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve_banded.html#scipy.linalg.solve_banded) I believe [ScaLAPACK does this as well](http://www.netlib.org/scalapack/slug/node85.html#SECTION04444000000000000000).This would mean storing A into a 3 * n matrix where the first row represents the superdiagonal (ignoring the first element of the first row), the second row represents the diagonal, and the third row represents the subdiagonal (ignoring the last element of the third row).  If this is done, I suggest changing the name of the argument from A to something else in order to avoid confusion. As you can see in the above link, Scipy uses ab instead. I personally think this is confusing as the vector b is not involved in the matrix representation. I suggest `Ad` for 'A diagonals'.

The proc definition would then look something like this:

```chapel
/* Return the solution ``x`` to the linear system ``A * x = b``.
    where ``A`` is a tridiagonal matrix.
    Note: "Ad" is the matrix "A" in matrix diagonal ordered form. 
*/
proc solve_tridiag (Ad: [?Adom] ?eltType, b: [?bdom] eltType) {
  // ...
}
```

The return type would be an array of elements with type `eltType` representing the solution vector x.

This is not the only way to design an interface for this function. So please do not hesitate to tell me if you can think of a better interface for this.
